src/idl_parser.cpp

```diff
@@ -1633,10 +1634,17 @@ CheckedError Parser::ParseEnum(bool is_union, EnumDef **dest) {
           if (ev.union_type.base_type != BASE_TYPE_STRUCT &&
               ev.union_type.base_type != BASE_TYPE_STRING)
             return Error("union value type may only be table/struct/string");
-          enum_def->uses_type_aliases = true;
         } else {
           ev.union_type = Type(BASE_TYPE_STRUCT, LookupCreateStruct(full_name));
         }
+        if (!enum_def->uses_ambiguous_type_aliases) {
```

Comment:
While this is a more lenient system, it does introduce the problem that the moment you add a second instance of a type in your union, all your code generation changes. Before, your code generation would change the moment you switch to using aliases, which is a little more predictable.
Not saying your approach is wrong, but we should be aware of this. It shouldn't affect too many people, but I wonder if it is worth it, being less predictable.

Reply:
Also the ident is a bit confusing, since it could sound like "type aliases" are ambiguous, whereas only the types are. So maybe `uses_multiple_type_instances` or something of that nature would be more informative.

Reply:
> While this is a more lenient system, it does introduce the problem that the moment you add a second instance of a type in your union, all your code generation changes. Before, your code generation would change the moment you switch to using aliases, which is a little more predictable.
> Not saying your approach is wrong, but we should be aware of this. It shouldn't affect too many people, but I wonder if it is worth it, being less predictable.

What do you think of removing the definition of `template<typename T> struct {{ENUM_NAME}}Traits` but always providing the declaration (regardless of `uses_multiple_type_instances`?  If `uses_multiple_type_instances` we can generate a comment or `static_assert` that explains why it's missing.  As a side effect it would also also make `{{ENUM_NAME}}Union::Set` stricter since it would report a compiler error instead of just setting `NONE` when given a type that is not in the union.  I'd do that in a separate PR, though.

Reply:
Other people may use those traits to map from type to enum? What do you mean by "always providing the declaration" (example)?

Reply:
> Other people may use those traits to map from type to enum? What do you mean by "always providing the declaration" (example)?

I don't think the default definition

```cpp
template<typename T> struct AnyTraits {
  static const Any enum_value = Any_NONE;
};
```

serves any purpose.  Rather it will hide errors since it allows me to call `AnyUnion::Set` with any type, silently discarding the value of any type that is not in the `Any` union.

I think we can just always declare it like this:

```cpp
template<typename T> struct AnyTraits;
```

So you'll get something like

```
error: incomplete type 'AnyTraits<int>' used in nested name specifier
```

There might be other ways (`static_assert` or such) to customize the error message but the main point highlight instead of swallowing what is most likely unintended behavior.

If possible we might want to customize the error message so in case the trait cannot be generated because of ambiguities you'll get a message telling you that.  Maybe that is too complex though and a comment in the generated code is sufficient.

Reply:
Ah ok, didn't understand. That does sound good, yes, though `_NONE` is also the union's null value, and it should be possible to produce such a value, which I think becomes impossible if we remove this case. Of course, having to use a random unused type to produce it is really bad, not sure if anyone relies on it.

Reply:
I think it'll still be possible.  I'll try to produce some code soonish.  That'll make it easier to talk.